server/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java

```diff
@@ -19,91 +19,31 @@
-import static java.time.temporal.ChronoField.HOUR_OF_DAY;
-import static java.time.temporal.ChronoField.MINUTE_OF_HOUR;
-import static java.time.temporal.ChronoField.MONTH_OF_YEAR;
-import static java.time.temporal.ChronoField.SECOND_OF_MINUTE;
-import static java.time.temporal.ChronoField.YEAR;
+import com.google.common.collect.Lists;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.common.SuppressLoggerChecks;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+
 
 /**
  * A logger that logs deprecation notices.
  */
 public class DeprecationLogger {
 
-    private final Logger logger;
-
-    /**
-     * This is set once by the {@code Node} constructor, but it uses {@link CopyOnWriteArraySet} to ensure that tests can run in parallel.
-     * <p>
-     * Integration tests will create separate nodes within the same classloader, thus leading to a shared, {@code static} state.
-     * In order for all tests to appropriately be handled, this must be able to remember <em>all</em> {@link ThreadContext}s that it is
-     * given in a thread safe manner.
-     * <p>
-     * For actual usage, multiple nodes do not share the same JVM and therefore this will only be set once in practice.
-     */
-    private static final CopyOnWriteArraySet<ThreadContext> THREAD_CONTEXT = new CopyOnWriteArraySet<>();
-
-    /**
-     * Set the {@link ThreadContext} used to add deprecation headers to network responses.
-     * <p>
-     * This is expected to <em>only</em> be invoked by the {@code Node}'s constructor (therefore once outside of tests).
-     *
-     * @param threadContext The thread context owned by the {@code ThreadPool} (and implicitly a {@code Node})
-     * @throws IllegalStateException if this {@code threadContext} has already been set
-     */
-    public static void setThreadContext(ThreadContext threadContext) {
-        Objects.requireNonNull(threadContext, "Cannot register a null ThreadContext");
-
-        // add returning false means it _did_ have it already
-        if (THREAD_CONTEXT.add(threadContext) == false) {
-            throw new IllegalStateException("Double-setting ThreadContext not allowed!");
-        }
-    }
-
-    /**
-     * Remove the {@link ThreadContext} used to add deprecation headers to network responses.
-     * <p>
-     * This is expected to <em>only</em> be invoked by the {@code Node}'s {@code close} method (therefore once outside of tests).
-     *
-     * @param threadContext The thread context owned by the {@code ThreadPool} (and implicitly a {@code Node})
-     * @throws IllegalStateException if this {@code threadContext} is unknown (and presumably already unset before)
-     */
-    public static void removeThreadContext(ThreadContext threadContext) {
-        assert threadContext != null;
+    private static final ThreadLocal<RingBuffer<String>> RECENT_WARNINGS = ThreadLocal.withInitial(() -> new RingBuffer<String>(20));
```

Comment:
This is only really for tests and won't work properly for integration tests. Currently we've only 2 tests that assert warnings, so not sure if it is worth to improve this now.